src/System.Net.Http/src/System/Net/Http/Unix/CurlHandler.ClientCertificateProvider.cs

```diff
@@ -83,20 +109,29 @@ private int TlsClientCertCallback(IntPtr ssl, out IntPtr certHandle, out IntPtr
                             if (!Interop.Ssl.SslAddExtraChainCert(sslHandle, dupCertHandle))
                             {
                                 EventSourceTrace("Failed to add extra chain certificate");
-                                return -1;
+                                return SuspendHandshake;
                             }
                         }
                     }
 
                     certHandle = _certHandle.DangerousGetHandle();
                     privateKeyHandle = _privateKeyHandle.DangerousGetHandle();
-                    return 1;
+
+                    EventSourceTrace("Client certificate set: {0}", certificate);
+                    return CertificateSet;
+                }
+                finally
+                {
+                    if (certificate != null && _clientCertificates != null) certificate.Dispose(); // only dispose cert if it's not newly created
```

Comment:
I don't understand this line... you're disposing the cert if it came from a user, but letting it get finalized if it was automatic?

Since the object came from the user, they might not expect that they're giving up the lifetime, so the user-provided cert shouldn't be disposed.  (They might be passing it on the next request)